jf_agent/git/__init__.py

```diff
@@ -0,0 +1,331 @@
+
+
+@dataclass
+class NormalizedUser:
+    id: str
+    name: str
+    login: str
+    email: str = None
+    url: str = None
+
+
+@dataclass
+class NormalizedBranch:
+    name: str
+    sha: str
+
+
+@dataclass
+class NormalizedProject:
+    id: str
+    name: str
+    login: str
+    url: str
+
+
+@dataclass
+class NormalizedShortRepository:
+    id: int
+    name: str
+    url: str
+
+
+@dataclass
+class NormalizedRepository:
+    id: int
+    name: str
+    full_name: str
+    url: str
+    is_fork: bool
+    default_branch_name: str
+    project: NormalizedProject
+    branches: List[NormalizedBranch]
+
+    def short(self):
+        # return the short form of Normalized Repository
+        return NormalizedShortRepository(id=self.id, name=self.name, url=self.url)
+
+
+@dataclass
+class NormalizedCommit:
+    hash: str
+    url: str
+    message: str
+    commit_date: str
+    author_date: str
+    author: NormalizedUser
+    repo: NormalizedShortRepository
+    is_merge: bool
+
+
+@dataclass
+class NormalizedPullRequestComment:
+    user: NormalizedUser
+    body: str
+    created_at: str
+
+
+@dataclass
+class NormalizedPullRequestReview:
+    user: NormalizedUser
+    foreign_id: int
+    review_state: str
+
+
+@dataclass
+class NormalizedPullRequest:
+    id: any
+    additions: int
+    deletions: int
+    changed_files: int
+    is_closed: bool
+    is_merged: bool
+    created_at: str
+    updated_at: str
+    merge_date: str
+    closed_date: str
+    title: str
+    body: str
+    url: str
+    base_branch: str
+    head_branch: str
+    author: NormalizedUser
+    merged_by: NormalizedUser
+    commits: List[NormalizedCommit]
+    comments: List[NormalizedPullRequestComment]
+    approvals: List[NormalizedPullRequestReview]
+    base_repo: NormalizedShortRepository
+    head_repo: NormalizedShortRepository
+
+
+class GitAdapter(ABC):
+    @abstractmethod
+    def get_users(self, include_projects) -> List[NormalizedUser]:
+        pass
+
+    @abstractmethod
+    def get_projects(self, include_projects, redact_names_and_urls) -> List[NormalizedProject]:
+        pass
+
+    @abstractmethod
+    def get_repos(
+        self, include_projects, include_repos, exclude_repos, redact_names_and_urls
+    ) -> List[NormalizedRepository]:
+        pass
+
+    @abstractmethod
+    def get_default_branch_commits(
+        self, api_repos, strip_text_content, server_git_instance_info, redact_names_and_urls
+    ) -> List[NormalizedCommit]:
+        pass
+
+    def get_pull_requests(
+        self, api_repos, strip_text_content, server_git_instance_info, redact_names_and_urls
+    ) -> List[NormalizedPullRequest]:
+        pass
+
+    def load_and_dump_git(self, config, endpoint_git_instance_info):
+        nrm_projects: List[NormalizedProject] = self.get_projects(
+            config.git_include_projects, config.git_redact_names_and_urls
+        )
+        write_file(config.outdir, 'bb_projects', config.compress_output_files, nrm_projects)
```

Comment:
This would be a good time to change the filename prefix on all of these from "bb_" to "git_". (And make the backend tolerate either convention.)